README.md
RemIdx
================

WARNING.....An issue #19 has been discovered, that currently leaves this tool useless.

Do NOT use it in it's currently state!

Check the ReadMe_First.pdf


changelog.txt
v0.0.1.4
- Remote Indexer:
-- Process will now run as low priority to prevent from bogging down the system.
-- Framework now to build win32 executable using PyInstaller.
-- ffmpeg's log level can now be set in the config file.
-- Allow default config values to be set by pressing enter.

V0.0.1.3
- Remote Indexer:
-- Issue 16 fixed: Bad bundle in workdir hangs RemIdx

V0.0.1.2
- Remote Indexer:
-- Added a configuration wizard

V0.0.1.1
- Removed start files, since they where not working.
- Updated doc with how-to start Remote Indexer
- PMS Agent:
-- Fixed issue with registration under TV-Shows
-- Corrected info and link in info.plist

V0.0.1.0
-- Now supported on Windows as well

V0.0.0.7
- Remote Indexer:
-- Added a Thread limit to FFMPEG

V0.0.0.6
- Remote Indexer:
-- Preparing to provide Windows support as well
- PMS Agent:
-- Preparing to provide Windows support as well

v0.0.0.5
- PMS Agent:
-- Can now add to indexes to database

v0.0.0.4
- Remote Indexer:
-- Added TV Shows
- PMS Agent:
-- Allow non-existing AspectRatio

v0.0.0.3
- Remote Indexer:
-- Added out directory
-- Added PMS Slam function
-- Added Allow PMS to fetch created bif files
- PMS Agent:
-- Added Slamming URL
-- Added Queue directory
-- Added Download bif files rutine

v0.0.0.2
- Added local workhorse
- Still not working

v0.0.0.1
- First public version
- Still not working



RemIdx.bundle/Contents/Code/__init__.py
import re
import os
from lxml import etree
import urllib2
import urllib

VERSION = ' V0.0.1.4'
NAME = L('RemIdx')
PREFIX = '/agents/remidx'
PLUGIN_NAME = 'remidx'
ART           = 'art-default.jpg'
ICON          = 'icon-default.png'
AGENTNAME = 'com.plexapp.agents.remidx'

####################################################################################################
def Start():
	print("********  Started %s on %s  **********" %(NAME  + VERSION, Platform.OS))
	Log.Debug("*******  Started %s on %s  ***********" %(NAME  + VERSION, Platform.OS))
	# Allow callback from remote indexer
	Plugin.AddPrefixHandler(PREFIX, Update, PLUGIN_NAME, ICON, ART)
	# Switch of auto indexing of libraries, sadly, framework only supports GET, so need to use urllib2
	opener = urllib2.build_opener(urllib2.HTTPHandler)
	request = urllib2.Request('http://127.0.0.1:32400/:/prefs?GenerateIndexFilesDuringAnalysis=0')
	request.get_method = lambda: 'PUT'
	url = opener.open(request)

###################################################################################################
# Movie agent
####################################################################################################		
class RemIdxMediaMovie(Agent.Movies):
	name = NAME + ' (Movies)'
	languages = [Locale.Language.NoLanguage]
	primary_provider = False
	contributes_to = ['com.plexapp.agents.imdb', 'com.plexapp.agents.themoviedb', 'com.plexapp.agents.none']
  	# Satisfy the framework here
	def search(self, results, media, lang):
		results.Append(MetadataSearchResult(id='null', score = 100))
    
	def update(self, metadata, media, lang, force):
		#Start to work
		GetMediaInfoMovie(media.id, media.title)

####################################################################################################
# TV Show agent
####################################################################################################		
class RemIdxMediaTV(Agent.TV_Shows):
	name = NAME + ' (TV)'
	primary_provider = False
	contributes_to = ['com.plexapp.agents.thetvdb', 'com.plexapp.agents.none']
  	# Satisfy the framework here
	def search(self, results, media, lang):
		results.Append(MetadataSearchResult(id='null', score = 100))

	def update(self, metadata, media, lang, force):
		GetMediaInfoTV(media.id, media.title)


####################################################################################################
# GetMediaInfo will grap some info for a movie media, and decide if futher action is needed
####################################################################################################		
@route(PREFIX + '/GetMediaInfoMovie')
def GetMediaInfoMovie(mediaID, myTitle):
	Log.Debug('Checking Movie media with an ID of : %s, and a title of : %s' %(mediaID, myTitle)) 
	myURL = 'http://' +  Prefs['This_PMS_IP'] + ':' + Prefs['This_PMS_Port']
	#Get the hash
	myNewURL = myURL + '/library/metadata/' + mediaID + '/tree'
	sections = XML.ElementFromURL(myNewURL).xpath('//MediaPart')
	for section in sections:
		myMediaHash = section.get('hash')
		Log.Debug('The hash for media %s is %s' %(mediaID, myMediaHash))
		# Does an index already exists?
		myIdxFile = os.path.join(Core.app_support_path, 'Media', 'localhost', myMediaHash[:1], myMediaHash[1:] + '.bundle', 'Contents', 'Indexes', 'index-sd.bif')
		Log.Debug('myIdxFile is : ' + myIdxFile)
		if os.path.isfile(myIdxFile):
			Log.Debug('Index exists for : %s with ID: %s, so skipping' %(myTitle, mediaID))
		else:
			Log.Debug('ID: %s - Index is missing for : %s' %(mediaID, myTitle))
			#Get media info
			myNewURL = myURL + '/library/metadata/' + mediaID
			# Grap the Section ID
			MediaContainer = XML.ElementFromURL(myNewURL)
			mySectionID = MediaContainer.get('librarySectionID')
			# Grap Media Info	
			sections = XML.ElementFromURL(myNewURL).xpath('//Media')
			for section in sections:
				myAspectRatio = section.get('aspectRatio')
				Log.Debug('Media AspectRatio for %s is %s' %(mediaID, myAspectRatio))
			#Get streaming info
			sections = XML.ElementFromURL(myNewURL).xpath('//Part')
			for section in sections:
				mySURL =  section.get('key')
			RegIdx(mySURL, myMediaHash, myTitle, mediaID, mySectionID, myAspectRatio)

####################################################################################################
# GetMediaInfo will grap some info for a TV-Show media, and decide if futher action is needed
####################################################################################################		
@route(PREFIX + '/GetMediaInfoTV')
def GetMediaInfoTV(mediaID, myTitle):
	Log.Debug('Checking TV media with an ID of : %s, and a title of : %s' %(mediaID, myTitle)) 
	myURL = 'http://' +  Prefs['This_PMS_IP'] + ':' + Prefs['This_PMS_Port']

	#Get mySectionID info
	myNewURL = myURL + '/library/metadata/' + mediaID
	MediaContainer = XML.ElementFromURL(myNewURL)
	mySectionID = MediaContainer.get('librarySectionID')
	Log.Debug('mySectionID is %s' %(mySectionID))

	# Get all children from this show
	myNewURL = myNewURL + '/allLeaves'
	Log.Debug('Inspecting URL: %s' %(myNewURL))
	episodes = XML.ElementFromURL(myNewURL).xpath('//Video')
	#Walk each episode
	for episode in episodes:
		myEpisodeTitle = myTitle + ' - ' + episode.get('title')
		Log.Debug('ID: %s Complete title is %s' %(mediaID, myEpisodeTitle))		
		mySURL = str(episode.xpath('Media/Part/@key'))[2:-2]
		Log.Debug('ID: %s Episode stream url is %s' %(mediaID, mySURL))
		myAspectRatio = str(episode.xpath('Media/@aspectRatio'))[2:-2]
		Log.Debug('ID: %s AspectRatio is %s' %(mediaID, myAspectRatio))
		#Get episode tree
		myKey = episode.get('ratingKey')
		myNewURL = myURL + '/library/metadata/' + myKey + '/tree'
		Log.Debug('ID: %s Tree URL is : %s' %(mediaID, myNewURL))
		myTree = XML.ElementFromURL(myNewURL).xpath('//MediaPart')
		for section in myTree:
			myMediaHash = section.get('hash')
			Log.Debug('ID: %s The hash for media %s is %s' %(mediaID, myKey, myMediaHash))
		# Does an index already exists?
		myIdxFile = os.path.join(Core.app_support_path, 'Media', 'localhost', myMediaHash[:1], myMediaHash[1:] + '.bundle', 'Contents', 'Indexes', 'index-sd.bif')
		Log.Debug('ID: %s myIdxFile is : %s' %(mediaID, myIdxFile))
		if os.path.isfile(myIdxFile):
			Log.Debug('Index exists for : %s with ID: %s, so skipping' %(myEpisodeTitle, myKey))
		else:
			Log.Debug('ID: %s Index is missing for : %s ' %(myKey, myEpisodeTitle))
			RegIdx(mySURL, myMediaHash, myEpisodeTitle, myKey, mySectionID, myAspectRatio)


####################################################################################################
# ReqIdx will request an index from the remote indexer
####################################################################################################		
@route(PREFIX + '/ReqIdx')
def RegIdx(mySURL, myMediaHash, myTitle, mediaID, mySectionID, myAspectRatio):
	myURL = 'http://' + Prefs['Remote_Idx_IP'] + ':' + Prefs['Remote_Port']+'/?Stream=http://' + Prefs['This_PMS_IP'] + ':' + Prefs['This_PMS_Port'] + mySURL + '&AspectRatio=' + myAspectRatio + '&SectionID=' + mySectionID + '&mediaID=' + mediaID + '&Title=' + String.Quote(myTitle) + '&Hash=' + myMediaHash + '.bundle'
	print NAME + ' is sending a request to remote Indexer for %s' %(myTitle)
	try:
		HTTP.Request(myURL, None, {'X-HTTP-Method-Override': 'QUEUE'}).content()
	except Exception:
		1
####################################################################################################
# Update function called from remote indexer, when a Bif file is ready
####################################################################################################		
@route(PREFIX + '/Update')
def Update():
	myURL = 'http://' + Prefs['Remote_Idx_IP'] + ':' + Prefs['Remote_Port'] + '/Out'
	try:
		Log.Debug('Got slammed by the Indexer')
		#Create a tmp storage directory in the Plug-In Support directory
		if not os.path.exists('Queue'):
			os.makedirs('Queue')

		request = urllib2.Request(myURL)
		# Set header so we can get access to directory listing of /Out on remote indexer
		request.add_header('X-HTTP-Method-Override', 'GETBIF')
		response = urllib2.urlopen(request)
		data = response.read()
		response.close()
		#TODO: Change this into XML		
		#Find string of bif files
		pos = data.find('<ul>') + 4
		pos2 = data.find('</ul>')
		data = data[pos:pos2]
		#Convert to list
		data3 = data.splitlines()
		#Skip first blank line
		data3.pop(0)
		for bif in data3:
			pos = bif.find('href="') + 6
			pos2 = bif.find('">')
			bif = bif[pos:pos2]
			Log.Debug('Got this file from the Remote Indexer: %s' %(bif))
			# Get the Media ID
			pos = bif.find('-')
			myMediaID = bif[:pos]
			Log.Debug('Media ID is : %s' %(myMediaID))
			sTargetDir = os.path.join(Core.app_support_path, 'Media', 'localhost', bif[pos+1:pos+2], os.path.splitext(bif[pos+2:])[0] + '.bundle', 'Contents', 'Indexes')
			Log.Debug('Target Directory is : %s' %(sTargetDir))
			# Create target dir if it doesn't exists
			if not os.path.exists(sTargetDir):
				os.makedirs(sTargetDir)
			myBifURL = myURL + '/' + bif
			Log.Debug('Download URL is %s' %(myBifURL))					
			#TODO: Need to check, if below was a success, before continue
			urllib.urlretrieve(myBifURL, sTargetDir + '/index-sd.bif')
			Add2Db(myMediaID)
			KillOnRemote(bif)

	except Exception:
		1
####################################################################################################
# This will add the newly placed index to the database
####################################################################################################
def Add2Db(myMediaID):
	opener = urllib2.build_opener(urllib2.HTTPHandler)
	#Enable Index Generation
	request = urllib2.Request('http://127.0.0.1:32400/:/prefs?GenerateIndexFilesDuringAnalysis=1')
	request.get_method = lambda: 'PUT'
	url = opener.open(request)
	#Analyze media
	request = urllib2.Request('http://127.0.0.1:32400/library/metadata/' + myMediaID + '/analyze')
	request.get_method = lambda: 'PUT'
	url = opener.open(request)
	#Disable Indexing
	request = urllib2.Request('http://127.0.0.1:32400/:/prefs?GenerateIndexFilesDuringAnalysis=0')
	request.get_method = lambda: 'PUT'
	url = opener.open(request)


####################################################################################################
# This will remove the entry from the remote indexers /Out directory
####################################################################################################
def KillOnRemote(bif):
	myURL = 'http://' + Prefs['Remote_Idx_IP'] + ':' + Prefs['Remote_Port']+'/?KillIt=' + bif
	try:
		HTTP.Request(myURL, None, {'X-HTTP-Method-Override': 'KILLOUT'}).content()
	except Exception:
		1

####################################################################################################
# Validate preferences
####################################################################################################
def ValidatePrefs():
	return

  
  


Remote_Indexer/RemIdx.py
#!/usr/bin/env python
#******************** Program Info **************************************
#
# This is the Remote Indexer workhorse of the RemIdx plug-in for Plex
# This must be running on a remote PC if possible, to
# avoid draining Plex Media Server for power
#
# Made by dane22 and srazer, Plex community members
# http://forums.plex.tv
#
# Code based on https://github.com/anachirino/bifserver, and she
# gracefully allowed me to steal it, and twist it into my needs
#
#***********************************************************************

# Search for TODO to find entry point for work in progress

VERSION = '0.0.1.4'

import logging
import io, json
import os
import sys
import shutil
import Queue
import threading
from urlparse import urlparse, parse_qs
import thread
import subprocess
import shlex
import time
import struct
import glob
import array
from SimpleHTTPServer import SimpleHTTPRequestHandler
from BaseHTTPServer import HTTPServer
import urllib2
import platform
import ConfigParser

# Initialize the Global vars used
PATH_TO_FFMPEG = ""
LOCAL_PORT = ""
FFMPEG_THREADS = ""
LOG_LEVEL = ""

# Enable or disable sending complete to server. Useful for devel.
SLAM = 1

# Loglevel for ffmpeg. Set empty string to load from config file. Valid values:
#	quiet:		Show nothing at all; be silent.
#	fatal:		Only show fatal errors. These are errors after which the process absolutely cannot continue after.
#	error:		Show all errors, including ones which can be recovered from.
#	warning:	Show all warnings and errors. Any message related to possibly incorrect or unexpected events will be shown.
#	info:		Show informative messages during processing. This is in addition to warnings and errors. This is the default value.
#	verbose:	Same as info, except more verbose.
#	debug:		Show everything, including debugging information.
FFMPEG_LOGLEVEL = "" 

# Set process to lowest priority
if sys.platform == 'win32':
	# Windows method
	import psutil
	pid = psutil.Process(os.getpid())
	pid.nice= psutil.IDLE_PRIORITY_CLASS
else:
	# Try the Linux/Mac method, keep on going if it fails.
	try: os.nice(19)
	except: pass

# Set window size. I do this mainly so when FFMPEG_LOGLEVEL is info or higher, it will display properly.
if sys.platform == 'win32':
	try:
		from ctypes import windll, byref, wintypes
		import time
		time.sleep(.5)
		width = 90
		height = 30
		buffer_height = 200
		hdl = windll.kernel32.GetStdHandle(-12)
#		os.system("mode con cols=" + str(width) + " lines=" + str(height)) # Kept here for reference
		rect = wintypes.SMALL_RECT(0, 50, 0+width-1, 50+height-1)  # (left, top, right, bottom)
		windll.kernel32.SetConsoleWindowInfo(hdl, True, byref(rect))
		time.sleep(.5) # Allow time for window size to change before changing buffer size.
		bufsize = wintypes._COORD(width, buffer_height) # columns, rows
		windll.kernel32.SetConsoleScreenBufferSize(hdl, bufsize)
		os.system("cls")
	# Don't panic if the above doesn't work.
	except: pass

#***********************************************************************
# Check FFMEGP location
#***********************************************************************
def CheckFFMPEG():
	if os.path.isfile(PATH_TO_FFMPEG):
		logging.debug('FFMPEG path validated as %s' %(PATH_TO_FFMPEG))
		return True
	else:
		print '***********************************************************************'
		print '*'
		print '*'
		print '*'
		print '*     ERROR!!!! Could not validate the path to the FFMPEG executable!!!'
		print '*     Check above message'
		print '*'
		print '*     Delete the file named RemIdx.ini to configure during next run'
		print '*'
		print '*     Quiting now!'
		print '*'
		print '*'
		print '***********************************************************************'
		logging.critical('FFMPEG not found at %s' %(PATH_TO_FFMPEG))
		return False

#***********************************************************************
# Shutdown Msg
#***********************************************************************
def ShutdownMsg(msg):
	print '***********************************************************************'
	print '* Currently working on item:'
	print '* ' + msg
	print '* Press <ENTER> to quit'
	print '***********************************************************************'

#***********************************************************************
# WebHandler
#***********************************************************************
class WebHandler(SimpleHTTPRequestHandler):
	def do_HEAD(self):
		self.get(False, 503)

	def do_GET(self):
		if self.headers.getheader('X-HTTP-Method-Override') == 'QUEUE':
			logging.debug('Got a Queue Req')
			self.get(False, 200)
		elif self.headers.getheader('X-HTTP-Method-Override') == 'GETBIF':
			logging.debug('Got a GetBif listing Req')
			SimpleHTTPRequestHandler.do_GET(self)
		elif self.headers.getheader('X-HTTP-Method-Override') == 'KILLOUT':
			logging.debug('Got a Kill Req')
			self.Kill_get(False, 200)
		else:
			if self.path.endswith(".bif"):
				logging.debug('Req to download bif file: ' + self.path)
				SimpleHTTPRequestHandler.do_GET(self)
			else:
				logging.critical('Got an invalid Req')
				self.get(True, 200)

	def Kill_get(self, include_body, queue_result):
		try:
			# We only hand requests for bif files..
			if urlparse(self.path).query.endswith(".bif"):
				dict = parse_qs(urlparse(self.path).query)
				self.sMyDir = os.path.realpath(os.path.dirname(sys.argv[0]))
				KillFile = os.path.join(self.sMyDir, 'Out', str(dict['KillIt'])[2:-2])
				logging.debug('About to remove file: ' + KillFile)
				os.remove(KillFile)
			else:
				logging.critical('Not a valid request')
				raise Exception('Not a valid request')
		except IOError:
			self.send_error(404, "Not Found")
			self.end_headers()
		except Exception as error:
			self.send_error(404, error.__str__())
			self.end_headers()
		finally:
			sys.stdout.flush()


	def get(self, include_body, queue_result):
		try:
			# We only hand requests for bif files..
			if urlparse(self.path).query.endswith(".bundle"):
				dict = parse_qs(urlparse(self.path).query)
				# Create Queue Entry
				self.sMyDir = os.path.realpath(os.path.dirname(sys.argv[0]))
				sQueueFile = os.path.join(self.sMyDir, 'Queue', str(dict['Hash'])[2:-2])
				sWorkFile = os.path.join(self.sMyDir, 'Work', str(dict['Hash'])[2:-2])
				if os.path.isfile(sWorkFile):
					#Already working on it
					logging.warning('The queue request recieved was already in progress')
					ShutdownMsg('The queue request recieved was already in progress')
				elif not os.path.isfile(sQueueFile):
					#Create an entry in the queue directory
					with io.open(sQueueFile, 'w', encoding='utf-8') as f:
						f.write(unicode(json.dumps(dict)))
			else:
				logging.critical('Not a valid request')
				raise Exception('Not a valid request')

		except IOError:
			self.send_error(404, "Not Found")
			self.end_headers()
		except Exception as error:
			self.send_error(404, error.__str__())
			self.end_headers()
		finally:
			sys.stdout.flush()

#***********************************************************************
# WebServer
#***********************************************************************
class WebServer(HTTPServer):
	def __init__(self):
		self.server_port = int(LOCAL_PORT)
		self.server_address = '0.0.0.0'
		self.HandlerClass = WebHandler
		self.server = HTTPServer((self.server_address, self.server_port), self.HandlerClass)
		print '* http deamon init okay'
		logging.debug('* http deamon init okay')

	def start(self):
		print '* Starting httpd deamon'	
		thread = threading.Thread(target = self.server.serve_forever)
		thread.deamon = True
		thread.start()
		sa = self.server.socket.getsockname()
		print "* Serving HTTP on", sa[0], "port", sa[1], "..."
		print '***********************************************************************'
		logging.debug('* Serving HTTP on %s port %s ....' %(sa[0],sa[1]))

	def stop(self):
		print 'Stopping httpd deamon.....Please wait'
		logging.info('Shutting down httpd deamon')
		self.server.shutdown()

#***********************************************************************
# Monitoring the Queue directory for work
#***********************************************************************
class BifQueue(threading.Thread):
	#Initialize the thread
	def __init__(self):
		threading.Thread.__init__(self)
		self.stopthread = threading.Event()
		logging.debug('BifQueue class initialized')
	#Running part
	def run(self):
		# Time between checking
		self.Interval = 5
		# Queue Directory
		self.myDir = os.path.realpath(os.path.dirname(sys.argv[0]))
		# Loop will continue until the stopthread event is set
		while not self.stopthread.isSet():
			myQList = glob.glob(os.path.join(self.myDir, 'Queue', '*.bundle'))
			if len(myQList) > 0:
				# Check if Work dir allready has an entry
				myWList = glob.glob(os.path.join(self.myDir, 'Work', '*.bundle'))
				if len(myWList) == 0:
					logging.debug('Need to move ' + myQList[0] + ' from Queue to Work')
					# Nothing to do, but work in queue
					sDestination = os.path.join(self.myDir, 'Work', os.path.basename(myQList[0]))
					# Move to work queue
					shutil.move(myQList[0],sDestination)
					GenJPGs(self.myDir)
			time.sleep(self.Interval)
	#Stop thread
	def stop(self):
		print 'Stopping BifQueue.....Please wait'
		logging.info('Shutting down BifQueue Thread')
		self.stopthread.set()

#***********************************************************************
# Generate the JPG files
#***********************************************************************
def GenJPGs(myDir):
	# Check if Work dir allready has an entry
	myWDir = os.path.join(myDir, 'Work', '*.bundle')
	myWList = glob.glob(myWDir)
	if len(myWList) > 0:
		#If tmp directory already existed, nuke it
		if os.path.exists(os.path.join(myDir, 'Tmp')):
			shutil.rmtree(os.path.join(myDir, 'Tmp'))
		#Create tmp directory
		os.makedirs(os.path.join(myDir, 'Tmp'))
		logging.debug('Started to work on bundle ' + myWList[0])
		try:
			#Extract needed info from work bundle json file
			with open(myWList[0]) as data_file:
				data = json.load(data_file)
				myHash = data["Hash"][0]
				logging.debug('Hash value is ' + myHash)
				myStream = data["Stream"][0]
				logging.debug('Stream URL is ' + myStream)
				myTitle = data["Title"][0]
				logging.debug('Title is ' + myTitle)
				mymediaID = data["mediaID"][0]
				logging.debug('MediaID is ' + mymediaID)
				mySectionID = data["SectionID"][0]
				logging.debug('Section ID is ' + mySectionID)
				try:
					myAspectRatio = data["AspectRatio"][0]
					logging.debug('myAspectRatio is ' + myAspectRatio)
				except:
					logging.debug('AspectRatio is missing')
					myAspectRatio = 'empty'
		except:
			# Bundlefile is invalid, so let's simply delete it
			logging.error('The bundle named %s was invalid, so deleting it' %(myWList[0]))
			os.remove(myWList[0])
			myTitle = '      ****** Idle, waiting for work ******'
			ShutdownMsg(myTitle)
			return

		#Tell my Master what's going on here
		print 'Starting to extract screenshots from %s' %myTitle
		print 'Close the window to terminate'
		# TODO start
		#Find the needed resolution for the jpg's
		if myAspectRatio > 1.5 :
			resolution = "240x136" # SD 16:9 ~ 1.7 ratio
		else :
			resolution = "240x180" # SD 4:3 ~ 1.3 ratio
		#Seems like Plex use this size regardless of media
		resolution = "320x136"
		# TODO end
		#Starting to grap screenshots	
		ffmpeg= [PATH_TO_FFMPEG, '-threads', FFMPEG_THREADS, '-loglevel', FFMPEG_LOGLEVEL, '-i', myStream, '-q', '3', '-s', resolution, '-r', '0.5', myDir + '/Tmp/%016d.jpg']
		if subprocess.call(ffmpeg):
			logging.critical('Could not extract images from video')
			raise Exception('Could not extract images from video')
			sys.exit(1)
		else:
			if MakeBIF(myDir, myHash, mymediaID, mySectionID, myStream):
				# Remove from work queue
				os.remove(myWList[0])
				slamPMS(myStream)
			else:
				logging.critical('MakeBif failed')
				print 'MakeBif failed !!!!!'
				sys.exit(1)
	else:
		myTitle = '      ****** Idle, waiting for work ******'
		ShutdownMsg(myTitle)	

#***********************************************************************
# Generate the bif file
#***********************************************************************
def MakeBIF(myDir, myHash, mymediaID, mySectionID, myStream):
	try:
		magic = [0x89,0x42,0x49,0x46,0x0d,0x0a,0x1a,0x0a]
		version = 0
		interval = 10
		myHash = os.path.splitext(myHash)[0]
		filename = os.path.join(myDir, 'Out', mymediaID + '-' + myHash + '.bif')
		#Create output directory
		if not os.path.exists(os.path.join(myDir, 'Out')):
			os.makedirs(os.path.join(myDir, 'Out'))
		files = os.listdir("%s" %(os.path.join(myDir, 'Tmp')))
		images = []
		for image in files:
			if image[-4:] == '.jpg':
				images.append(image)
		images.sort()
		images = images[1:]

		f = open(filename, "wb")
		array.array('B', magic).tofile(f)
		f.write(struct.pack("<I1", version))
		f.write(struct.pack("<I1", len(images)))
		f.write(struct.pack("<I1", 1000 * interval))
		array.array('B', [0x00 for x in xrange(20,64)]).tofile(f)

		bifTableSize = 8 + (8 * len(images))
		imageIndex = 64 + bifTableSize
		timestamp = 0

		# Get the length of each image
		for image in images:
			statinfo = os.stat("%s/%s" % (os.path.join(myDir, 'Tmp'), image))
			f.write(struct.pack("<I1", timestamp))
			f.write(struct.pack("<I1", imageIndex))

			timestamp += 1
			imageIndex += statinfo.st_size

		f.write(struct.pack("<I1", 0xffffffff))
		f.write(struct.pack("<I1", imageIndex))

		# Now copy the images
		for image in images:
			data = open("%s/%s" % (os.path.join(myDir, 'Tmp'), image), "rb").read()
			f.write(data)

		f.close()
		#Remove Tmp files
		shutil.rmtree(os.path.join(myDir, 'Tmp'))

		myTitle = '****** Idle, waiting for work ******'
		ShutdownMsg(myTitle)
		logging.debug('Bif file created for media hashed %s' %(myHash))
		return True
	except:
		logging.critical('Error in creating a bif for file with a hash of %s' %(myHash))
		return False

#***********************************************************************
# Show Banner
#***********************************************************************
def ShowBanner():
	print ''
	print ''
	print ''
	print ''
	print ''
	print ''
	print '***********************************************************************'
	print '* Welcome to the RemIdx Remote Indexer version ' + VERSION
	print '*'
	print '* Running on ' + platform.system()
	print '*'
	print '* Made by dane22 and srazer, Plex community members'
	print '*'
	print '* To reconfigure, either edit the file named RemIdx.ini, or delete it'
	print '*'
	print '***********************************************************************'
	print '* Path to ffmpeg is: ' + PATH_TO_FFMPEG
	print '***********************************************************************'

#***********************************************************************
# Logger
#***********************************************************************
def Logging():
	# Configure logging
	if LOG_LEVEL.upper() == 'NONE':
		logging.basicConfig(filename=None)
	else:		
		log_level = getattr(logging, LOG_LEVEL.upper(), None)		
		if not isinstance(log_level, int):
	    		raise ValueError('Invalid log level: %s' % log_level)
		logging.basicConfig(filename='RemIdx.log',level=log_level,format='%(asctime)s - %(levelname)s - %(funcName)s - %(lineno)d - %(message)s')

#***********************************************************************
# Slam PMS about a Bif-file is ready
#***********************************************************************
def slamPMS(myStream):
	#TODO: This one cause an error on the remote indexer!!!!
	# We sadly always ends up with the exception
	pos = myStream.find('/library/parts')
	PMSURL = myStream[:pos] + '/agents/remidx'
	print ('Slamming PMS @ : %s' %(PMSURL))
	logging.debug('Slamming PMS @ : %s' %(PMSURL))
	#Sending Slam
	request = urllib2.Request(PMSURL)
	try:
		if SLAM:
			response = urllib2.urlopen(request, timeout=60)
			response.close()
	except:
		print 'Slamming okay'
	# Tell my Master I'm falling asleep
	myTitle = '          ****** Idle, waiting for work ******'
	ShutdownMsg(myTitle)

#***********************************************************************
# conf function
#***********************************************************************
class conf():
	def __init__(self, sMyDir):
		if not os.path.isfile(os.path.join(sMyDir,'RemIdx.ini')):
			# No ini file, so let's create one
			with open(os.path.join(sMyDir,'RemIdx.ini'), 'wb') as configfile:
				self.Config = ConfigParser.ConfigParser()
				self.Config.read(os.path.join(sMyDir,'RemIdx.ini'))
				self.Config.add_section('Configuration')
				self.Config.add_section('RemIdx')
			self.SetConf(sMyDir)
		else:
			# Found an ini file			
			self.Config = ConfigParser.ConfigParser()
			self.Config.read(os.path.join(sMyDir,'RemIdx.ini'))
			try:
				if not self.Config.getboolean("Configuration", "IsSet"):
					self.SetConf(sMyDir)
				else:
					self.ReadConf()
			except:
				# During last configuration, my master aborted :-(
				with open(os.path.join(sMyDir,'RemIdx.ini'), 'wb') as configfile:
					self.Config = ConfigParser.ConfigParser()
					self.Config.read(os.path.join(sMyDir,'RemIdx.ini'))
					self.Config.add_section('Configuration')
					self.Config.add_section('RemIdx')
				self.SetConf(sMyDir)

	def ReadConf(self):
		# FFMPEG Path
		global PATH_TO_FFMPEG
		PATH_TO_FFMPEG = self.Config.get('RemIdx', 'PATH_TO_FFMPEG')
		# Port I'm listing to
		global LOCAL_PORT
		LOCAL_PORT = str(self.Config.getint('RemIdx', 'LOCAL_PORT'))
		global FFMPEG_THREADS
		FFMPEG_THREADS = str(self.Config.get('RemIdx', 'FFMPEG_THREADS'))
		global LOG_LEVEL
		LOG_LEVEL = self.Config.get('RemIdx', 'LOG_LEVEL')
		global FFMPEG_LOGLEVEL
		# If FFMPEG_LOGLEVEL is unset then load it from the config file.
		if FFMPEG_LOGLEVEL == '': FFMPEG_LOGLEVEL = self.Config.get('RemIdx', 'FFMPEG_LOGLEVEL')
		# If FFMPEG_LOGLEVEL is still unset then set a value.
		if FFMPEG_LOGLEVEL == '': FFMPEG_LOGLEVEL = 'quiet'


	def SetConf(self, sMyDir):
		print ''
		print ''
		print ''
		print ''
		print ''
		print ''
		print '***********************************************************************'
		print '* Welcome to the RemIdx Remote Indexer version ' + VERSION
		print '*'
		print '* Running on ' + platform.system()
		print '*'
		print '* Made by dane22 and srazer, Plex community members'
		print '*'
		print '* This application needs to be configured first' 
		global PATH_TO_FFMPEG
		while not os.path.isfile(PATH_TO_FFMPEG):
			print '*'
			print '* Please enter the full path to the FFMPEG executable, and press <ENTER>'
			PATH_TO_FFMPEG = raw_input('ffmpeg path: ')
		self.Config.set('RemIdx', 'PATH_TO_FFMPEG', PATH_TO_FFMPEG)
		print '*'
		print '* Thanks.....'
		global LOCAL_PORT
		while not ((LOCAL_PORT>32400) and (LOCAL_PORT<32501)):
			print '*'
			print '* Now please enter the port number this application shall be listnening on'
			print '* Remember to open your firewall for this port, and I recommend port 32405'
			print '* Valid range is 32401 to 32500'
			LOCAL_PORT = raw_input('Remote Indexer Port [32405]: ')
			# Allow user to set default by just pressing enter.
			if LOCAL_PORT == '': LOCAL_PORT = 32405
			LOCAL_PORT = int(LOCAL_PORT)
		LOCAL_PORT = str(LOCAL_PORT)
		self.Config.set('RemIdx', 'LOCAL_PORT', LOCAL_PORT)
		print '*'
		print '* Thanks.....'
		VALID_FFMPEG_THREADS = ('auto', '1', '2', '3', '4', '5', '6', '7', '8')
		global FFMPEG_THREADS
		while not (FFMPEG_THREADS in VALID_FFMPEG_THREADS):
			print '*'
			print '* Now please enter the amount of CPU cores to use (type auto to use all cores)'
			print '* If this is a dedicated Indexer, I recommend you type auto'
			print '* Valid options are ' + str(VALID_FFMPEG_THREADS)[1:-1]
			FFMPEG_THREADS = raw_input('Number of cores [auto]: ')
			# Allow user to set default by just pressing enter.
			if FFMPEG_THREADS == '': FFMPEG_THREADS = 'auto'
		self.Config.set('RemIdx', 'FFMPEG_THREADS', FFMPEG_THREADS)
		print '*'
		print '* Thanks.....'
		global LOG_LEVEL
		VALID_LOG_LEVEL = ('none', 'debug', 'info', 'warning', 'error', 'critical')
		while not (LOG_LEVEL in VALID_LOG_LEVEL):
			print '*'
			print '* Now please enter the log level to use'
			print '* Valid levels are ' + str(VALID_LOG_LEVEL)[1:-1]
			LOG_LEVEL = raw_input('Log Level [info]: ')
			# Allow user to set default by just pressing enter.
			if LOG_LEVEL == '': LOG_LEVEL = 'info'
		self.Config.set('RemIdx', 'LOG_LEVEL', LOG_LEVEL)
		self.Config.set('RemIdx', 'FFMPEG_LOGLEVEL', 'quiet')

		# Writing our configuration file to 'RemIdx.ini'
		self.Config.set('Configuration', 'isset', True)
		with open(os.path.join(sMyDir,'RemIdx.ini'), 'wb') as configfile:
			self.Config.write(configfile)

#***********************************************************************
# Main function
#***********************************************************************
def main():
	# Directory I live in
	sMyDir = os.path.realpath(os.path.dirname(sys.argv[0]))
	conf(sMyDir)
	# Configure logging
	Logging()
	logging.info('***********************************************************************')
	logging.info('Starting RemIdx Indexer version %s on %s' %(VERSION, platform.system()))
	# Show banner
	ShowBanner()
	# Check for correct FFMPEG Path
	if not CheckFFMPEG():
		print 'Quitting'
		logging.info('Quitting RemIdx Indexer')
		sys.exit(1)
	logging.debug('Starting from directory named : ' + sMyDir)
	# Create Queue and Work Directory
	myQDir = os.path.join(sMyDir,'Queue')
	if not os.path.exists(myQDir):
		os.makedirs(myQDir)
	myWDir = os.path.join(sMyDir,'Work')
	if not os.path.exists(myWDir):
		os.makedirs(myWDir)
	# Start WebServer
	myWeb = WebServer()
	myWeb.start()
	# Start Queue monitoring
	myQueue = BifQueue()
	myQueue.deamon = True
	myQueue.start()
	# Check if some work is already waiting for us to pick up
	GenJPGs(sMyDir)
	#Wait for my Master is pressing <ENTER>
	raw_input('')	
	myQueue.stop()
	myWeb.stop()
	logging.info('Quitting RemIdx Indexer')
	print 'Quitting.....Please wait'
	sys.exit(0)

if __name__ == '__main__':
	main()

